File: msglib-java/src/msglib/proto/MSerializer.java
package msglib.proto;

import java.lang.reflect.Array;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.TreeMap;

import msglib.annotation.MsgField;
import msglib.proto.io.IDataBuffer;
import msglib.proto.io.MBinaryProto;
import msglib.proto.io.MException;
import msglib.proto.io.MProto;
import msglib.proto.io.MTextProto;
import msglib.proto.io.MemoryDataBuffer;
import msglib.proto.io.MemoryTextBuffer;
import msglib.proto.types.MField;
import msglib.proto.types.MList;
import msglib.proto.types.MMap;
import msglib.proto.types.MStruct;
import msglib.proto.types.MType;

public class MSerializer {

    private static MSerializer instance = new MSerializer();

    public static MSerializer inst() {
        return instance;
    }

    public static String SerializeAsText(Object data) throws MException {
        MemoryTextBuffer buffer = new MemoryTextBuffer();
        MProto target = new MTextProto(buffer);
        instance.Write(target, data);
        return buffer.getText();
    }

    public static <T> T Deserialize(String text, Class<T> clazz) throws MException {
        MemoryTextBuffer buffer = new MemoryTextBuffer(text);
        MProto source = new MTextProto(buffer);
        return (T) instance.Read(source, clazz);
    }

    public static byte[] Serialize(Object data) throws MException {
        MemoryDataBuffer buffer = new MemoryDataBuffer();
        MProto target = new MBinaryProto(buffer);
        instance.Write(target, data);
        return buffer.toArray();
    }

    public static void Serialize(Object msg, IDataBuffer buffer) throws MException {
        MProto target = new MBinaryProto(buffer);
        instance.Write(target, msg);
    }

    public static <T> T Deserialize(IDataBuffer buffer, Class<T> clazz) throws MException {
        MProto source = new MBinaryProto(buffer);
        return (T) instance.Read(source, clazz);
    }

    public static <T> T Deserialize(byte[] data, Class<T> clazz) throws MException {
        MemoryDataBuffer buffer = new MemoryDataBuffer(data, false);
        MProto source = new MBinaryProto(buffer);
        return (T) instance.Read(source, clazz);
    }

    public static <T> T Deserialize(byte[] data, int offset, int length, Class<T> clazz) throws MException {
        MemoryDataBuffer buffer = new MemoryDataBuffer(data, offset, length, false);
        MProto source = new MBinaryProto(buffer);
        return (T) instance.Read(source, clazz);
    }

    public static <T> T DecodeNoThrow(IDataBuffer buffer, Class<T> clazz) {
        try {
            return Deserialize(buffer, clazz);
        }
        catch (MException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static <T> T DecodeNoThrow(byte[] data, Class<T> clazz) {
        try {
            return Deserialize(data, clazz);
        }
        catch (MException e) {
            e.printStackTrace();
            return null;
        }
    }

    public void Write(MProto target, Object data) throws MException {
        Write(target, data, null);
    }

    public void Write(MProto target, Object data, Class<?> type) throws MException {
        Write(target, data, type, null);
    }

    public void Write(MProto target, Object data, Class<?> type, MMeta meta) throws MException {
        if (type == null)
            type = data.getClass();
        if (meta == null)
            meta = MMeta.GetClassMeta(type);

        if (meta == null || MType.Null == meta.Type) {
            throw new MException("Unsupported Type: " + type);
        }
        if (MType.Bool == meta.Type) {
            target.WriteBool((Boolean) data);
            return;
        }
        if (MType.Byte == meta.Type) {
            target.WriteByte((Byte) data);
            return;
        }
        if (MType.I16 == meta.Type) {
            target.WriteI16((Short) data);
            return;
        }
        if (MType.I32 == meta.Type) {
            target.WriteI32((Integer) data);
            return;
        }
        if (MType.I64 == meta.Type) {
            target.WriteI64((Long) data);
            return;
        }
        if (MType.Float == meta.Type) {
            target.WriteFloat((Float) data);
            return;
        }
        if (MType.Double == meta.Type) {
            target.WriteDouble((Double) data);
            return;
        }
        if (MType.Binary == meta.Type) {
            if (meta.MainType.isArray()) {
                target.WriteBinary((byte[]) data);
                return;
            }
            throw new MException("Binary should be byte[]");
        }
        if (MType.String == meta.Type) {
            target.WriteString((String) data);
            return;
        }
        if (MType.Struct == meta.Type) {
            MStruct struc = new MStruct();
            struc.Name = type.getName();
            Map<Integer, Field> fields = GetStructFields(type);
            target.WriteStructBegin(struc);
            //
            for (Entry<Integer, Field> entry : fields.entrySet()) {
                Field fieldinfo = entry.getValue();
                if (null != fieldinfo.getAnnotation(Deprecated.class))
                    continue;
                MMeta fieldmeta = MMeta.GetFieldMeta(fieldinfo);
                MField field = new MField(fieldinfo.getName(), fieldmeta.Type, entry.getKey());
                //
                Object value = null;
                try {
                    value = fieldinfo.get(data);
                }
                catch (Exception e) {
                    throw new MException("Field Access Error: " + type.getName() + "." + fieldinfo.getName(), e);
                }
                if (value == null)
                    continue;
                target.WriteFieldBegin(field);
                Write(target, value, fieldinfo.getType(), fieldmeta);
            }
            target.WriteFieldStop();

            return;
        }

        if (MType.Map == meta.Type) {
            MMeta keymeta = MMeta.GetClassMeta(meta.KeyType);
            MMeta valmeta = MMeta.GetClassMeta(meta.ValueType);
            Map dict = (Map) data;
            MMap map = new MMap(keymeta.Type, valmeta.Type, dict.size());
            target.WriteMapBegin(map);
            for (Object obj : dict.entrySet()) {
                Entry entry = (Entry) obj;
                Object key = entry.getKey();
                Object value = entry.getValue();
                Write(target, key, meta.KeyType, keymeta);
                Write(target, value, meta.ValueType, valmeta);
            }
            return;
        }

        if (MType.List == meta.Type) {
            MMeta valmeta = MMeta.GetClassMeta(meta.ValueType);
            if (meta.MainType.isArray()) {
                int count = Array.getLength(data);
                //
                MList list = new MList(valmeta.Type, count);
                target.WriteListBegin(list);
                for (int i = 0; i < count; i++) {
                    Object entry = Array.get(data, i);
                    Write(target, entry, meta.ValueType, valmeta);
                }
                return;
            }

            if (List.class.isAssignableFrom(meta.MainType)) {
                List arr = (List) data;
                MList list = new MList(valmeta.Type, arr.size());
                target.WriteListBegin(list);
                for (Object entry : arr) {
                    Write(target, entry, meta.ValueType, valmeta);
                }
                return;
            }

            throw new MException("unsupported list type: " + meta.MainType.getName());
        }

        throw new MException("Unsupported Type: " + meta.Type);
    }

    /**
     * TODO: [optimization] cache field info of given type
     * @param type
     * @return
     * @throws MException
     */
    private Map<Integer, Field> GetStructFields(Class<?> type) throws MException {
        Map<Integer, Field> fields = new TreeMap<Integer, Field>();
        for (Field field : type.getFields()) {
            MsgField msgfield = field.getAnnotation(MsgField.class);
            if (msgfield == null) {
                continue;
            }
            if (fields.containsKey(msgfield.Id())) {
                throw new MException("Duplicate FieldId, struct = " + type.getCanonicalName() + ", field = "
                        + msgfield.Id() + ", name = " + field.getName());
            }
            fields.put(msgfield.Id(), field);
        }
        return fields;
    }

    public Object Read(MProto source, Class<?> type) throws MException {
        return Read(source, type, null);
    }

    public Object Read(MProto source, Class<?> type, MMeta meta) throws MException {
        if (meta == null)
            meta = MMeta.GetClassMeta(type);
        if (meta == null || MType.Null == meta.Type) {
            throw new MException("Unsupported Type: " + type);
        }
        else if (MType.Bool == meta.Type) {
            return source.ReadBool();
        }
        else if (MType.Byte == meta.Type) {
            return source.ReadByte();
        }
        else if (MType.I16 == meta.Type) {
            return source.ReadI16();
        }
        else if (MType.I32 == meta.Type) {
            return source.ReadI32();
        }
        else if (MType.I64 == meta.Type) {
            return source.ReadI64();
        }
        else if (MType.Float == meta.Type) {
            return source.ReadFloat();
        }
        else if (MType.Double == meta.Type) {
            return source.ReadDouble();
        }
        else if (MType.Binary == meta.Type) {
            byte[] bytes = source.ReadBinary();
            if (meta.MainType.isArray()) {
                return bytes;
            }
            else {
                throw new MException("Binary shoud be byte[]");
            }
        }
        else if (MType.String == meta.Type) {
            return source.ReadString();
        }
        else if (MType.Struct == meta.Type) {
            MStruct struc = source.ReadStructBegin();
            Object data = null;
            try {
                data = type.getConstructor().newInstance();
            }
            catch (Exception e) {
                throw new MException("Error to Create Struct " + type.getName(), e);
            }

            Map<Integer, Field> fields = GetStructFields(type);
            while (true) {
                MField field = source.ReadFieldBegin();
                if (field.Type == MType.Null)
                    break;
                if (fields.containsKey(field.ID)) {
                    Field prop = fields.get(field.ID);
                    MMeta propmeta = MMeta.GetFieldMeta(prop);
                    Object value = Read(source, prop.getType(), propmeta);
                    try {
                        prop.set(data, value);
                    }
                    catch (Exception e) {
                        throw new MException("Error Writing Field " + type.getName() + "." + prop.getName(), e);
                    }
                }
                else {
                    Skip(source, field.Type);
                }
            }
            return data;
        }
        else if (MType.Map == meta.Type) {
            MMap map = source.ReadMapBegin();

            if (!Map.class.isAssignableFrom(type))
                throw new MException("Map type should be assignable to java.util.Map");

            Object data = null;
            if (type.isInterface()) {
                data = new HashMap();
            }
            else {
                try {
                    data = type.getConstructor().newInstance();
                }
                catch (Exception e) {
                    throw new MException("Failed to create Map " + type.getName(), e);
                }
            }

            Map dict = (Map) data;
            for (int i = 0; i < map.Count; i++) {
                Object key = Read(source, meta.KeyType);
                Object value = Read(source, meta.ValueType);
                dict.put(key, value);
            }

            return data;
        }
        else if (MType.List == meta.Type) {
            MList list = source.ReadListBegin();
            Object data = null;
            if (type.isArray()) {
                data = Array.newInstance(meta.ValueType, list.Count);
                for (int i = 0; i < list.Count; i++) {
                    Object value = Read(source, meta.ValueType);
                    Array.set(data, i, value);
                }
            }
            else {
                if (!List.class.isAssignableFrom(type))
                    throw new MException("List type should be assignable to java.util.List");

                if (type.isInterface()) {
                    data = new ArrayList();
                }
                else {
                    try {
                        data = type.getConstructor().newInstance();
                    }
                    catch (Exception e) {
                        throw new MException("Error Creating List " + type.getName(), e);
                    }
                }

                List arr = (List) data;
                for (int i = 0; i < list.Count; i++) {
                    Object value = Read(source, meta.ValueType);
                    arr.add(value);
                }
            }
            return data;
        }
        else {
            throw new MException("Unsupported MsgType " + type);
        }
    }

    private void Skip(MProto source, int type) throws MException {
        if (MType.Null == type) {
            throw new MException("Unsupported Type: " + type);
        }
        else if (MType.Bool == type) {
            source.ReadBool();
        }
        else if (MType.Byte == type) {
            source.ReadByte();
        }
        else if (MType.I16 == type) {
            source.ReadI16();
        }
        else if (MType.I32 == type) {
            source.ReadI32();
        }
        else if (MType.I64 == type) {
            source.ReadI64();
        }
        else if (MType.Float == type) {
            source.ReadFloat();
        }
        else if (MType.Double == type) {
            source.ReadDouble();
        }
        else if (MType.Binary == type) {
            source.ReadBinary();
        }
        else if (MType.String == type) {
            source.ReadString();
        }
        else if (MType.Struct == type) {
            MStruct struc = source.ReadStructBegin();
            while (true) {
                MField field = source.ReadFieldBegin();
                if (field.Type == MType.Null)
                    break;
                Skip(source, field.Type);
            }
        }
        else if (MType.Map == type) {
            MMap map = source.ReadMapBegin();
            for (int i = 0; i < map.Count; i++) {
                Skip(source, map.KeyType);
                Skip(source, map.ValueType);
            }
        }
        else if (MType.List == type) {
            MList list = source.ReadListBegin();
            for (int i = 0; i < list.Count; i++) {
                Skip(source, list.ElementType);
            }
        }
        else {
            throw new MException("Unsupported MsgType " + type);
        }
    }

}

File: msglib-csharp/MsgLib/Serialize/MSerializer.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Reflection;

namespace MsgLib
{
    internal class MSerializer
    {
        private static MSerializer instance = null;
        internal static MSerializer Default 
        {
            get
            {
                if (instance == null)
                {
                    instance = new MSerializer();
                }
                return instance;
            }
        }

        internal T Deserialize<T>(byte[] bytes)
        {
            MemoryStream ms = new MemoryStream(bytes, false);
            return Deserialize<T>(ms);
        }

        internal T Deserialize<T>(MemoryStream ms)
        {
            MemoryDataBuffer buffer = new MemoryDataBuffer(ms);
            MProto proto = new MBinaryProto(buffer);
            object res = Read(proto, typeof(T));
            return (T)res;
        }

        internal void Serialize(MemoryStream stream, object data)
        {
            MemoryDataBuffer buffer = new MemoryDataBuffer(stream);
            MProto proto = new MBinaryProto(buffer);
            Write(proto, data);
        }
		internal void Write(MProto target, object data)
		{
			Write(target, data, null, null);
		}
		internal void Write(MProto target, object data, Type type)
		{
			Write(target, data, type, null);
		}
        internal void Write(MProto target, object data, Type type , MMeta meta )
        {
            if (type == null)
                type = data.GetType();
            if (meta == null)
                meta = MMeta.GetMMeta(type);
            if (meta == null || MType.Null == meta.MType)
            {
                throw new MException("Unsupported Type: " + type);
            }
            else if (MType.Bool == meta.MType)
            {
                target.WriteBool((bool)data);
            }
            else if (MType.Byte == meta.MType)
            {
                target.WriteByte((sbyte)data);
            }
            else if (MType.I16 == meta.MType)
            {
                target.WriteI16((short)data);
            }
            else if (MType.I32 == meta.MType)
            {
                target.WriteI32((int)data);
            }
            else if (MType.I64 == meta.MType)
            {
                target.WriteI64((long)data);
            }
            else if (MType.Float == meta.MType)
            {
                target.WriteFloat((float)data);
            }
            else if (MType.Double == meta.MType)
            {
                target.WriteDouble((double)data);
            }
            else if (MType.Binary == meta.MType)
            {
                if (meta.MainType.IsArray)
                {
                    target.WriteBinary((byte[])data);
                }
                else
                {
                    var seg = (ArraySegment<byte>)data;
                    target.WriteBinary(seg.Array, seg.Offset, seg.Count);
                }
            }
            else if (MType.String == meta.MType)
            {
                target.WriteString((string)data);
            }
            else if (MType.Struct == meta.MType)
            {
                MStruct struc = new MStruct();
                struc.Name = type.Name;
                var fields = GetStructFields(type);
                target.WriteStructBegin(struc);
                // 
                foreach (var entry in fields)
                {
                    var proptype = entry.Value.PropertyType;
                    var fieldmeta = MMeta.GetMMeta(proptype);
                    var value = entry.Value.GetValue(data, null);
                    if (value == null)
                        continue;
                    var field = new MField(entry.Value.Name, fieldmeta.MType, entry.Key);
                    target.WriteFieldBegin(field);
                    Write(target, value, proptype, fieldmeta);
                }
                target.WriteFieldStop();
            }
            else if (MType.Map == meta.MType)
            {
                var keymeta = MMeta.GetMMeta(meta.KeyType);
                var valmeta = MMeta.GetMMeta(meta.ValueType);
                var dict = data as IDictionary;
                var map = new MMap(keymeta.MType, valmeta.MType, dict.Count);
                target.WriteMapBegin(map);
                foreach (var key in dict.Keys)
                {
                    Write(target, key, meta.KeyType, keymeta);
                    var value = dict[key];
                    Write(target, value, meta.ValueType, valmeta);
                }
            }
            else if (MType.List == meta.MType)
            {
                var valmeta = MMeta.GetMMeta(meta.ValueType);
                var arr = data as IEnumerable;
                int count = 0;
                if (meta.MainType.IsArray)
                {
                    count = (data as Array).Length;
                }
                else
                {
                    count = (data as IList).Count;
                }
                //
                var list = new MList(valmeta.MType, count);
                target.WriteListBegin(list);
                foreach (var entry in arr)
                {
                    Write(target, entry, meta.ValueType, valmeta);
                }
            }
            else if (MType.Set == meta.MType)
            {
                var valmeta = MMeta.GetMMeta(meta.ValueType);
                int count = (int)meta.ContainerType.GetProperty("Count").GetValue(data, null);
                var arr = data as IEnumerable; 
                var list = new MList(valmeta.MType, count);
                target.WriteListBegin(list);
                foreach (var entry in arr)
                {
                    Write(target, entry, meta.ValueType, valmeta);
                }
            }
        }

        private IDictionary<int, PropertyInfo> GetStructFields(Type type)
        {
            var fields = new SortedDictionary<int, PropertyInfo>();
            var allprops = type.GetTypeProperties();
            foreach (var prop in allprops)
            {
                var attrs = prop.GetCustomAttributes(typeof(MFieldAttribute), false);
                if (attrs.Length == 0)
                {
                    continue;
                }
                int id = (attrs[0] as MFieldAttribute).ID;
                fields.Add(id, prop);
            }
            return fields;
        }
		internal object Read(MProto source, Type type)
		{
			return Read(source, type, null);
		}
        internal object Read(MProto source, Type type, MMeta meta )
        {
            if (meta == null)
                meta = MMeta.GetMMeta(type);
            if (meta == null || MType.Null == meta.MType)
            {
                throw new MException("Unsupported Type: " + type);
            }
            else if (MType.Bool == meta.MType)
            {
                return source.ReadBool();
            }
            else if (MType.Byte == meta.MType)
            {
                return source.ReadByte();
            }
            else if (MType.I16 == meta.MType)
            {
                return source.ReadI16();
            }
            else if (MType.I32 == meta.MType)
            {
                return source.ReadI32();
            }
            else if (MType.I64 == meta.MType)
            {
                return source.ReadI64();
            }
            else if (MType.Float == meta.MType)
            {
                return source.ReadFloat();
            }
            else if (MType.Double == meta.MType)
            {
                return source.ReadDouble();
            }
            else if (MType.Binary == meta.MType)
            {
                byte[] bytes = source.ReadBinary();
                if (meta.MainType.IsArray)
                {
                    return bytes;
                }
                else
                {
                    return new ArraySegment<byte>(bytes);
                }
            }
            else if (MType.String == meta.MType)
            {
                return source.ReadString();
            }
            else if (MType.Struct == meta.MType)
            {
                MStruct struc = source.ReadStructBegin();
                var data = Activator.CreateInstance(type, true);
                var fields = GetStructFields(type);
                while (true)
                {
                    MField field = source.ReadFieldBegin();
                    if (field.Type == MType.Null)
                        break;
                    if (fields.ContainsKey(field.ID))
                    {
                        var prop = fields[field.ID];
                        var propmeta = MMeta.GetMMeta(prop.PropertyType);
						if (propmeta.MType != field.Type)
						{
							throw new MException("FieldErr, field = " + type.Name + ":"
                                + prop.Name + ", got type = " + field.Type);
						}
						else
						{
	                        var val = Read(source, prop.PropertyType, propmeta);
	                        prop.SetValue(data, val, null);
						}
                    }
                    else
                    {
                        Skip(source, field.Type);
                    }
                }
                return data;
            }
            else if (MType.Map == meta.MType)
            {
                var map = source.ReadMapBegin();
                object data = null;
                if (type.IsInterface)
                {
                    var instType = typeof(Dictionary<,>);
                    instType = instType.MakeGenericType(meta.KeyType, meta.ValueType);
                    data = Activator.CreateInstance(instType);
                }
                else
                {
                    data = Activator.CreateInstance(type);
                }
                var dict = data as IDictionary;
                for (int i = 0; i < map.Count; i++)
                {
                    var key = Read(source, meta.KeyType);
                    var val = Read(source, meta.ValueType);
                    dict.Add(key, val);
                }
                return data;
            }
            else if (MType.List == meta.MType)
            {
                var list = source.ReadListBegin();
                object data = null;
                if (type.IsArray)
                {
                    data = Array.CreateInstance(meta.ValueType, list.Count);
                    var arr = data as Array;
                    for (int i = 0; i < list.Count; i++)
                    {
                        var value = Read(source, meta.ValueType);
                        arr.SetValue(value, i);
                    }
                }
                else
                {
                    if (type.IsInterface)
                    {
                        var instType = typeof(List<>);
                        instType = instType.MakeGenericType(meta.ValueType);
                        data = Activator.CreateInstance(instType);
                    }
                    else
                    {
                        data = Activator.CreateInstance(type);
                    }
                    var arr = data as IList;
                    for (int i = 0; i < list.Count; i++)
                    {
                        var value = Read(source, meta.ValueType);
                        arr.Add(value);
                    }
                }
                return data;
            }
            else if (MType.Set == meta.MType)
            {
                var list = source.ReadListBegin();
                object data = null;
                var instType = type;
                if (type.IsInterface)
                {
                    instType = typeof(HashSet<>).MakeGenericType(meta.ValueType);
                }
                else if (type.IsGenericTypeDefinition)
                {
                    instType = type.MakeGenericType(meta.ValueType);
                }
                data = Activator.CreateInstance(instType);
                var addMethod = type.GetMethod("Add");
                for (int i = 0; i < list.Count; i++)
                {
                    var value = Read(source, meta.ValueType);
                    addMethod.Invoke(data, new object[] { value });
                }
                return data;
            }
            throw new MException("Should Not Reach Here");
        }

        private void Skip(MProto source, MType mType)
        {
            if (MType.Null == mType)
            {
                throw new MException("Unsupported Type: " + mType);
            }
            else if (MType.Bool == mType)
            {
                source.ReadBool();
            }
            else if (MType.Byte == mType)
            {
                source.ReadByte();
            }
            else if (MType.I16 == mType)
            {
                source.ReadI16();
            }
            else if (MType.I32 == mType)
            {
                source.ReadI32();
            }
            else if (MType.I64 == mType)
            {
                source.ReadI64();
            }
            else if (MType.Float == mType)
            {
                source.ReadFloat();
            }
            else if (MType.Double == mType)
            {
                source.ReadDouble();
            }
            else if (MType.Binary == mType)
            {
                source.ReadBinary();
            }
            else if (MType.String == mType)
            {
                source.ReadString();
            }
            else if (MType.Struct == mType)
            {
                MStruct struc = source.ReadStructBegin();
                while (true)
                {
                    MField field = source.ReadFieldBegin();
                    if (field.Type == MType.Null)
                        break;
                    Skip(source, field.Type);
                }
            }
            else if (MType.Map == mType)
            {
                var map = source.ReadMapBegin();
                for (int i = 0; i < map.Count; i++)
                {
                    Skip(source, map.KeyType);
                    Skip(source, map.ValueType);
                }
            }
            else if (MType.List == mType)
            {
                var list = source.ReadListBegin();
                for (int i = 0; i < list.Count; i++)
                {
                    Skip(source, list.ElementType);
                }
            }
            else if (MType.Set == mType)
            {
                var list = source.ReadListBegin();
                for (int i = 0; i < list.Count; i++)
                {
                    Skip(source, list.ElementType);
                }
            }
            else
            {
                throw new MException("Should Not Reach Here, type = " + mType);
            }
        }
    }

    internal class MMeta
    {
        internal MType MType { get; set; }

        internal Type MainType { get; set; }
        internal Type KeyType { get; set; }
        internal Type ValueType { get; set; }
        internal Type ContainerType { get; set; } // IDictonary or IList
		
		internal MMeta() : this (MType.Null)
		{
		}
        internal MMeta(MType type )
        {
            this.MType = type;
        }

        internal static MMeta GetMMeta(Type type)
        {
            MMeta meta = new MMeta();
            meta.MainType = type;
            if (type == typeof(String))
            {
                meta.MType = MType.String;
            }
            else if (type == typeof(Boolean))
            {
                meta.MType = MType.Bool;
            }
            else if (type == typeof(Byte) || type == typeof(SByte))
            {
                meta.MType = MType.Byte;
            }
            else if (type == typeof(Int16) || type == typeof(UInt16))
            {
                meta.MType = MType.I16;
            }
            else if (type == typeof(Int32) || type == typeof(UInt32))
            {
                meta.MType = MType.I32;
            }
            else if (type == typeof(Int64) || type == typeof(UInt64))
            {
                meta.MType = MType.I64;
            }
            else if (type == typeof(Single))
            {
                meta.MType = MType.Float;
            }
            else if (type == typeof(Double))
            {
                meta.MType = MType.Double;
            }
            else if (type == typeof(byte[]) || type == typeof(ArraySegment<byte>))
            {
                meta.MType = MType.Binary;
            }
            else if (type.IsArray)
            {
                if (type.GetArrayRank() == 1)
                {
                    meta.MType = MType.List;
                    var elemType = type.GetElementType();
                    meta.ValueType = elemType;
                }
            }
            else if (type.IsGenericType)
            {
                if (!FillGenericeType(meta, type))
                {
                    foreach (var itype in type.GetInterfaces())
                    {
                        if (FillGenericeType(meta, itype))
                            break;
                    }
                }
            }
            else
            {
                var structAttrs = type.GetCustomAttributes(typeof(MStructAttribute), false);
                if (structAttrs.Length > 0)
                {
                    meta.MType = MType.Struct;
                }
            }

            return meta;
        }

        private static bool FillGenericeType(MMeta meta, Type itype)
        {
            if (!itype.IsGenericType)
                return false;
            if (!itype.IsGenericType)
                return false;
            var gtype = itype.GetGenericTypeDefinition();
            var targs = itype.GetGenericArguments();
            if (gtype == typeof(IDictionary<,>))
            {
                gtype = gtype.MakeGenericType(targs);
                meta.MType = MType.Map;
                meta.KeyType = targs[0];
                meta.ValueType = targs[1];
                meta.ContainerType = gtype;
                return true;
            }
            else if (gtype == typeof(IList<>))
            {
                gtype = gtype.MakeGenericType(targs);
                meta.MType = MType.List;
                meta.ValueType = targs[0];
                meta.ContainerType = gtype;
                return true;
            }
            return false;
        }
    }

}
